aten/src/ATen/native/cuda/ForeachReduceOp.cu

```diff
@@ -445,54 +257,95 @@ std::vector<Tensor> foreach_tensor_norm_cuda(
   const auto options = tensors[0].options();
   const ScalarType output_dtype =
       dtype.has_value() ? dtype.value() : tensors[0].scalar_type();
-  const ScalarType output_per_tensor_dtype = toOpMathType(output_dtype);
-  auto output_per_tensor = at::zeros(
-      {static_cast<int64_t>(ntensors) * max_chunks_per_tensor},
-      options.dtype(output_per_tensor_dtype));
+  const ScalarType opmath_output_dtype = toOpMathType(output_dtype);
+  const auto opmath_t_options = options.dtype(opmath_output_dtype);
+
+  Tensor output_per_tensor;
+  if (calculate_norm_per_tensor) {
+    output_per_tensor = at::zeros(
+        {static_cast<int64_t>(ntensors) * max_chunks_per_tensor},
+        opmath_t_options);
+  }
+  Tensor global_output;
+  if (calculate_global_norm) {
+    global_output = at::zeros({320}, opmath_t_options);
+  }
+
+  auto tensor_lists = std::vector<std::vector<Tensor>>{tensors.vec()};
 
-  std::vector<at::Tensor> vec_res;
-  vec_res.reserve(ntensors);
   const auto res_option = options.dtype(output_dtype);
-  for (const auto i : c10::irange(ntensors)) {
-    vec_res.push_back(at::native::empty_cuda(
+  std::vector<at::Tensor> vec_res;
+  if (calculate_norm_per_tensor) {
+    vec_res.reserve(ntensors);
+    for (const auto i : c10::irange(ntensors)) {
+      vec_res.push_back(at::native::empty_cuda(
+          {},
+          optTypeMetaToScalarType(res_option.dtype_opt()),
+          res_option.layout_opt(),
+          res_option.device_opt(),
+          res_option.pinned_memory_opt(),
+          res_option.memory_format_opt()));
+    }
+  }
+  Tensor global_norm;
+  if (calculate_global_norm) {
+    global_norm = at::native::empty_cuda(
         {},
         optTypeMetaToScalarType(res_option.dtype_opt()),
         res_option.layout_opt(),
         res_option.device_opt(),
         res_option.pinned_memory_opt(),
-        res_option.memory_format_opt()));
+        res_option.memory_format_opt());
   }
 
-  auto tensor_lists = std::vector<std::vector<Tensor>>{tensors.vec()};
-
   AT_DISPATCH_FLOATING_TYPES_AND2(
       kHalf,
       c10::kBFloat16,
       tensor_lists[0][0].scalar_type(),
-      "foreach_tensor_norm_cuda_scalar_type",
+      "foreach_tensor_norm_cuda_impl_scalar_type",
       [&]() {
-        // using opmath_t = typename at::opmath_type<scalar_t>;
         AT_DISPATCH_OUT_DTYPES(
-            output_dtype, "foreach_tensor_norm_cuda_out_dtype", [&]() {
+            output_dtype, "foreach_tensor_norm_cuda_impl_out_dtype", [&]() {
               using out_opmath_t = typename at::opmath_type<out_t>;
               if (p == static_cast<double>(1)) {
                 multi_tensor_apply<1>(
                     tensor_lists,
                     LpNormFunctor<scalar_t, NormType::L1, out_t>(),
-                    output_per_tensor.mutable_data_ptr<out_opmath_t>(),
-                    max_chunks_per_tensor);
+                    calculate_global_norm
```

Comment:
these ternary's would look clumsy but basically I'm trying to avoid creating buffers if they aren't needed, i.e., foreach_global_norm doesn't need buffers for per-tensor norm and foreach_norm, one for global norm